lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -0,0 +1,291 @@
+      initTruncatedInput = rewriter.create<linalg::InitTensorOp>(
+          loc, expandedInputType.getShape(), rewriter.getIntegerType(32));
+      updatesTensor = rewriter.create<linalg::InitTensorOp>(
+          loc, inputType.getShape(), resultElemType);
+    } else {
+      Value inputSize = rewriter.create<tensor::DimOp>(loc, input, 0);
+      initTruncatedInput = rewriter.create<linalg::InitTensorOp>(
+          loc, ValueRange{inputSize}, expandedInputType.getShape(),
+          rewriter.getIntegerType(32));
+      updatesTensor = rewriter.create<linalg::InitTensorOp>(
+          loc, ValueRange{inputSize}, resultElemType);
+    }
+    Value constantOne = rewriter.create<arith::ConstantIntOp>(
+        loc, 1, resultElemType.getIntOrFloatBitWidth());
+    updatesTensor =
+        rewriter.create<linalg::FillOp>(loc, constantOne, updatesTensor)
+            .getResult(0);
+
+    // Converting the input element type from i64 to i32.
+    AffineMap indexingMap =
+        AffineMap::getMultiDimIdentityMap(2, op->getContext());
+    indexingMaps.clear();
+    indexingMaps.insert(indexingMaps.begin(), {indexingMap, indexingMap});
+    iteratorTypes[0] = getParallelIteratorTypeName();
+
+    Value truncatedInput =
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, initTruncatedInput.getType(),
+                /*input*/ expandedInput,
+                /*output*/ initTruncatedInput,
+                /*indexingMaps=*/indexingMaps,
+                /*iteratorTypes=*/iteratorTypes,
+                [&](OpBuilder &b, Location loc, ValueRange args) {
+                  Value trunc = b.create<arith::TruncIOp>(
+                      loc, b.getIntegerType(32), args[0]);
+                  b.create<linalg::YieldOp>(loc, trunc);
+                })
+            ->getResult(0);
+
+    Value bincountSize =
+        rewriter.create<arith::MaxSIOp>(loc, maxInput, minlength);
+    bincountSize = castIntToIndex(rewriter, loc, bincountSize);
+    Value bincountTensor = createInitTensor(rewriter, loc, {bincountSize},
+                                            resultElemType, constantZero);
+
+    auto scatterOp = rewriter.create<TMTensor::ScatterOp>(
+        loc, bincountTensor.getType(),
+        ValueRange{updatesTensor, truncatedInput}, ValueRange{bincountTensor},
+        /*unique_indices*/ false);
+
+    Region &scatterOpRegion = scatterOp.region();
+    auto &scatterOpBlock = scatterOpRegion.emplaceBlock();
+    scatterOpBlock.addArgument(resultElemType, scatterOpRegion.getLoc());
+    scatterOpBlock.addArgument(resultElemType, scatterOpRegion.getLoc());
+    auto blockArgs = scatterOpBlock.getArguments();
+
+    // Creating an add instruction inside the scatter op region to increment the
+    // frequency counter with one.
+    OpBuilder regionBuilder(scatterOpRegion);
+    Value add = regionBuilder.create<arith::AddIOp>(scatterOp->getLoc(),
+                                                    blockArgs[0], blockArgs[1]);
```

Comment:
this can be replaced with `blockArgs[1] + 1` and the update tensor is not needed. also, please change the `blockArgs[0]` and `blockArgs[1]` to `update` and `bincount` for readability.